### ISSUE TITLE
EarlyStopping() does not return the model weights corresponding to the epoch with the best value of the monitored quantity

### ISSUE THREAD

**Issue by GeorgeOrfanidis** [keras-team/keras#20323]
When using `early_stopping = keras.callbacks.EarlyStopping(monitor="loss", mode="auto", patience=50, verbose=1, restore_best_weights=True)` with `restore_best_weights=True` the restored weights do not correspond to the epoch that produced the best value of the monitored quantity but rather to the epoch after the best epoch. 

For a demonstration of the aforementioned observation, consider the log: 

1/1 ━━━━━━━━━━━━━━━━━━━━ 2s 2s/step - error_metric: 3.1894 - loss: -5.8280 
Epoch 2/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 55ms/step - error_metric: 3.8508 - loss: -1.1713 
.
.
.
**omitted for clarity**
.
.
.
Epoch 22/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 137ms/step - error_metric: 0.2297 - loss: -15.9473
Epoch 23/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 161ms/step - error_metric: 0.8144 - loss: -15.3367
Epoch 24/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 126ms/step - error_metric: 0.6884 - loss: -15.5262
**Epoch 25/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 135ms/step - error_metric: 0.0093 - loss: -15.9999**
Epoch 26/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 122ms/step - error_metric: 0.6519 - loss: -15.5750
Epoch 27/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 99ms/step - error_metric: 0.6260 - loss: -15.6082
.
.
.
**omitted for clarity**
.
.
.
Epoch 75/6000
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 32ms/step - error_metric: 0.0595 - loss: -15.9965
Epoch 75: early stopping
Restoring model weights from the end of the best epoch: **25.**

Evidently, the epoch produced the min loss is epoch 25 with measured loss = -15.9999; however, when using ` model.predict(X)` (note: X is exactly the same validation set) the produced loss = -15.5750 which is the loss measured on epoch 26, the epoch after the best epoch. 

It seems `keras.callbacks.EarlyStopping()` restores the model weights from the epoch following the best epoch!

**Comment by mehtamansi29** [comment]
Hi @GeorgeOrfanidis -

Thanks for reporting the issue. Can you please provide the sample code to reproduce this issue ?

**Comment by GeorgeOrfanidis** [comment]
Hello @mehtamansi29 and thank you for replying to my issue! While I can not share the original code because it's from an unpublished work, I have created an example that reproduces the issue.

~~~
import numpy as np
import keras
import tensorflow as tf

np.random.seed(42)
X = tf.constant(3.45 * np.array([1.,2.,3.,4.,5.,6.,7.]) + (np.random.randn(7) * 0.1),dtype=tf.float32)
X = tf.reshape(X, (1,7,1))

Y = tf.constant(3.45 * np.array([1.,2.,3.,4.,5.,6.,7.]),dtype=tf.float32)
Y = tf.reshape(Y, (1,7,1))

def scheduler(epoch, lr):
    if epoch < 30:
      return lr
    else:
      lr = float(1e-3)
      return  lr

callback_lr = keras.callbacks.LearningRateScheduler(scheduler)

input = keras.layers.Input(shape=(X.shape[1],))
output = keras.layers.Dense(7,activation="linear", kernel_initializer=keras.initializers.GlorotUniform(seed=42))(input)
model = keras.Model(inputs=[input], outputs=[output])
model.summary()
model.compile(loss="mse", optimizer=keras.optimizers.Adam(learning_rate=1e-1))
early_stopping = keras.callbacks.EarlyStopping(monitor="loss", mode="auto", patience=2, verbose=1, restore_best_weights=True)
history = model.fit(X,Y, epochs=6000, callbacks=[early_stopping, callback_lr])
~~~

**Output:**
![image](https://github.com/user-attachments/assets/007c2038-de6a-4039-8772-3204b4a3a2d5)

~~~
Y_pred = model.predict(X)

Y_pred = tf.convert_to_tensor(Y_pred)
Y_true = tf.reshape(Y,(1,7))

loss_fn = keras.losses.MeanSquaredError() 
loss = loss_fn(Y_true, Y_pred)
print(loss)
~~~

**Output:**
![image](https://github.com/user-attachments/assets/9d45786e-773f-4d18-a823-62962a0e2448)


Observing the log and the output of the `EarlyStopping()` callback, the epoch that achieved the min loss is epoch 6 with a loss 33.2164 however, when using the `model.predict()` function with the "best weights" on the same input the loss achieved is 38.8818 which corresponds to the loss of the epoch following the best epoch (epoch 7 as it can be seen from log).

**Comment by mehtamansi29** [comment]
Hi @GeorgeOrfanidis -

Thanks for providing sample code. Early stopping is required to stop training when monitor metric stops improving. 

Here as per your code,`EarlyStopping()` callback is monitor loss metric  and because of  `restore_best_weights=True` argument, saving the weights after epoch6 because loss is very least(33.2164). 'patience=2' argument is there in `EarlyStopping()` , so after epoch6 loss is increasing for epoch7 and epoch8. 
Due to this reason, for reducing the overfitting early stopping triggers after epoch8 as there is no such improvement in loss in consecutive 2 epochs('patience=2'). 

More details on `EarlyStopping()` you can find [here](https://keras.io/api/callbacks/early_stopping/).

**Comment by GeorgeOrfanidis** [comment]
Hello @mehtamansi29, 

As you mentioned in your response, `EarlyStopping()` with `restore_best_weights=True` is supposed to save the weights corresponding to the epoch that achieved the min value of the monitored metric, which in this case is epoch6 with loss= 33.2164. When using the model and applying the same exact input (as demonstrated in my code above) the model should achieve the same exact loss with epoch6 however, the model achieves the loss of the epoch7, indicating that the stored wights do not correspond to the ones after the end of epoch6 but rather to the ones of the next epoch, epoch7. 

Note: to demonstrate the bug I am training and using the model on the exact same instance.

**Comment by mehtamansi29** [comment]
Hi @GeorgeOrfanidis -

Here your sample code output is like this.
```
Epoch 1/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 2s 2s/step - loss: 270.5954 - learning_rate: 0.1000
Epoch 2/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 273ms/step - loss: 113.6899 - learning_rate: 0.1000
Epoch 3/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 248ms/step - loss: 58.7962 - learning_rate: 0.1000
Epoch 4/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 145ms/step - loss: 45.0612 - learning_rate: 0.1000
Epoch 5/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 148ms/step - loss: 36.8283 - learning_rate: 0.1000
Epoch 6/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 132ms/step - loss: 33.2164 - learning_rate: 0.1000
Epoch 7/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 144ms/step - loss: 38.8818 - learning_rate: 0.1000
Epoch 8/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 167ms/step - loss: 48.5781 - learning_rate: 0.1000
Epoch 8: early stopping
Restoring model weights from the end of the best epoch: 6.
```

Here getting best loss metric  at this line 
```
Epoch 6/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 132ms/step - loss: 33.2164 - learning_rate: 0.1000
```
And due to `restore_best_weights=True` arguments getting best weights when loss metric is least, which is at epoch6 not epoch7. 
Epoch7 loss metric is this:
```
Epoch 7/50
1/1 ━━━━━━━━━━━━━━━━━━━━ 0s 144ms/step - loss: 38.8818 - learning_rate: 0.1000
```

**Comment by GeorgeOrfanidis** [comment]
Hello @mehtamansi29, 

I totally understand this however, when you use the returned model with `Y_pred = model.predict(X)` and `X` being exactly the same instance as in during training the returned loss is 38.8818 which is the loss corresponding to epoch 7 and not the loss corresponding to the best epoch, epoch 6 in that case. This indicates that the presumably best weights are not from the best epoch (epoch 6 in this case) but rather the next epoch (epoch 7).

**Comment by mehtamansi29** [comment]
Hi @GeorgeOrfanidis -

> when you use the returned model with `Y_pred = model.predict(X)` and `X` being exactly the same instance as in during training the returned loss is 38.8818 which is the loss corresponding to epoch 7 and not the loss corresponding to the best epoch

While predicting the model not getting expected result as per training restored weights(epoch6), this is because sometime training has some non-persistent weights so weights get overwrite and maybe  due to  patience parameter in `EarlyStopping()`
callbacks.

`restore_best_weights: Whether to restore model weights from the epoch with the best value of the monitored quantity. If False, the model weights obtained at the last step of training are used. An epoch will be restored regardless of the performance relative to the baseline. If no epoch improves on baseline, training will run for patience epochs and restore weights from the best epoch in that set. Defaults to False.`

In your code, changing `patience=1`, in this line `early_stopping = keras.callbacks.EarlyStopping(monitor="loss", mode="auto", patience=1, verbose=1, restore_best_weights=True)` will give correct loss while doing `Y_pred = model.predict(X)`. Same thing you can try to in your original code as well. 

Alternative approach you can manually save weights using mode.save_weights() and model.load_weights() for accurate result. 

Attached [gist](https://colab.sandbox.google.com/gist/mehtamansi29/306e6aea012789e25b0d72179aef5a9c/20323-earlystopping-does-not-return-the-model-weights-corresponding-to-the-epoch-with-the-best-value-of-the-monitored-quantity.ipynb) for your reference.

**Comment by GeorgeOrfanidis** [comment]
Hi @mehtamansi29, 

I tried your suggestion, using `patience=1`, and I am still getting the loss corresponding to the epoch after the best epoch, the loss of epoch 7. Note: The same happens in the gist you shared above. When using `model.predict(X)`, the calculated loss is 38.881756 (loss of epoch 7 during training) and not the best loss achieved during epoch 6. This observation indicates that the `patience` parameter does not fix the issue. 

> Alternative approach you can manually save weights using mode.save_weights() and model.load_weights() for accurate result. 

This is what I am currently doing; however, `EarlyStopping()` is widely used, and it seems it does not work as described in the documentation. It does not return the weights corresponding to the epoch that achieved the best value for the monitored metric but rather the weights of the next epoch. It seems there is a problem with how `self.best_weights` variable is updated in [https://github.com/keras-team/keras/blob/v3.6.0/keras/src/callbacks/early_stopping.py#L10](url).

**Comment by GeorgeOrfanidis** [comment]
Hi @mehtamansi29, 

I hope you're doing well! I wanted to follow up and ask if you had any thoughts or suggestions on my previous comment. I would greatly appreciate your insights.

**Comment by fchollet** [comment]
The behavior you are seeing is completely normal... it's a result of the weirdness of:

1. Using "loss" as the monitor. Never do this -- it's meaningless. The loss will always go down over time.
2. Setting 1 epoch == 1 batch.

The statement "Restoring model weights from the end of the best epoch: 6." is 100% correct. That's what happens. And the loss obtained by that model is 38.8818. That's because when you run batch/epoch #6, the loss for the batch corresponds to the state of the model at the *start* of the batch. While running the batch, the model gets updated. The model state at the end of the batch is the post-update one, and that's the one we save, and it's the one that gets 38.8818 (which is printed when you compute the loss on the next batch).

This goes away if you use validation data (which you should ALWAYS do with EarlyStopping) since validation metrics are compute at the end of each epoch, on the state of the model as it was at the end of the epoch. It always goes away if you use normal epochs with more than 1 batch per epoch.
